enhancements/microshift/loadbalancer-service-support.md

```diff
@@ -0,0 +1,201 @@
+what CNI is used by the cluster, and behave differently based on the CNI type.
+
+### Risks and Mitigations
+
+The pluggable CNI design of Microshift is still under discussion, we need to
+make sure that CNI type can be aware in Microshift binary.
+
+### Drawbacks
+
+It is not a CNI agnostic design. The loadbalancer Service controller will behave
+differently between OVN-Kubernetes and other CNIs.
+
+## Design Details
+
+The Service of loadblancer type is designed to exposes the Service externally
+using a cloud provider's load balancer. Normally, there is a loadbalancer
+Service controller which is responsible for:
+
+1. provision a loadbalancer for the Service.
+2. publish the ingress endpoint information in the Service's
+   `.status.loadBalancer` field.
+
+However, in the context of Microshift, as we only get one node in the cluster,
+it doesn't make sense to use a real load balancer to forward the ingress
+traffic. So instead of provisioning the loadbalancer instance, the controller
+is only responsible for helping the CNIs to plumbing the ingress traffic.
+
+### Traffic Plumbing
+
+#### OVN-Kubernetes
+
+OVN-Kubernetes has already implemented the [loadbalancer ingress
+IP](https://github.com/openshift/ovn-kubernetes/blob/master/docs/external-ip-and-loadbalancer-ingress.md)
+support. The ovnkube-node component which is running on each host watches the
+Service's `.status.loadBalancer` field, and insert iptables rules that does DNAT
+to the clusterIP of the Service. Therefore, from the loadbalancer Service
+controller perspective, it will only be responsible for the updating the status
+of the Service. The ovn-kubernetes will do the rest of the job.
+
+#### Other CNIs
+
+For other CNIs, which don't have the similar function as OVN-Kubernetes, neither
+do they bypass the kernel for the ingress traffic, we can take the similar
+approach as K3S. The K3S's loadbalancer Service controller provisions a
+[klipper-lb](https://github.com/k3s-io/klipper-lb) pod to each nodes for each
+loadbalancer Service. The pod can insert iptables rules which DNAT ingress
+traffic to the nodeIP of the Service.
+
+Another option is manipulating the iptables rule by this controller itself.
+Since there is only one node, and the microshsift binary has the access to the
+host network.
+
+### Update Status
+As, we don't want to ask user to allocated additional VIP as the loadbalancer
+IP. We can say that the ingress ip of the Service's `.status.loadBalancer` field
+shall always be the existing IPs of the node. When there are multiple IPs, the
+kubelet `NodeIP` shall be used by default.
+
+The controller shall also check if the Service port has already been allocated
+to other loadbalancer Service before updating the Service status.
+
+### Open Questions [optional]
+
+* Do we want to allow users to choose the IP other than the NodeIP as the
+Ingress IP of loadbalancer Service, when there are multiple IPs available?
```

Comment:
Is that a feature supported in OCP?

Reply:
That is a feature of OVN-K. OVN-K doesn't care if the IP is used as the LB service. It will just use whatever IP which is set in the `.status.loadBalancer`.

Reply:
This isn't my area of expertise, so I'm not sure what reasons a user would have for picking an IP other than the node IP. Do we have that requirement? Does saying no, that we only want to support the node IP, make it harder to change our mind about that later?

Reply:
This is similar to https://issues.redhat.com/browse/NP-621.

Reply:
I asked about OCP support because we want to think carefully about supporting something in MicroShift that is not exposed in OCP.

Do we need to answer this question to move ahead? Or can we leave this as an open question for now and wait for an RFE to help us shape requirements for setting the IP?

Reply:
As for now, OCP doesn't have this feature, as we would rely on a real loadbalancer like metal-lb to expose VIPs instead of the IPs of nodes. I think we can keep this as an open question for now.